File: StudyGamePrograming04maze/NavComponent.cpp
#include "NavComponent.h"
#include "Tile.h"

NavComponent::NavComponent(class Actor* owner, int updateOrder)
	: MoveComponent(owner, updateOrder)
	, mNextNode(nullptr)
{}

void NavComponent::Update(float deltaTime)
{
	if (mNextNode)
	{
		Vector2 diff = mNextNode->GetPosition() - mOwner->GetPosition();
		float nextspeed = speed;
		if (speed * deltaTime > diff.Length()) { nextspeed = diff.Length() / deltaTime; }
		if (diff.Length() <= speed * deltaTime)
		{
			mNextNode = mNextNode->GetParent();
		}
		mNextDirVector = Vector2::Normalize(diff);
		mNextDirAngle = Math::Atan2(-mNextDirVector.y, mNextDirVector.x);
		SetVelocity(nextspeed * mNextDirVector);		
	}

	MoveComponent::Update(deltaTime);
}

void NavComponent::SetStartNode(class Tile* start)
{
	if (start) {
		mNextNode = start->GetParent();
	}
}

File: StudyGamePrograming04maze_VB/NavComponent.vb
Imports System.Numerics

Public Class NavComponent
    Inherits MoveComponent
    Sub New(ByRef owner As Actor, ByVal updateOrder As Integer)
        MyBase.New(owner, updateOrder)
        mNextNode = Nothing
    End Sub
    Public Overrides Sub Update(deltaTime As Double)
        If (mNextNode IsNot Nothing) Then
            Dim diff As Vector2 = mNextNode.GetPosition() - mOwner.GetPosition()
            Dim nextspeed = speed
            If (speed * deltaTime > diff.Length) Then nextspeed = Int(diff.Length) / deltaTime
            If (diff.Length <= speed * deltaTime) Then
                mNextNode = mNextNode.GetParent()
            End If
            mNextDirVector = Vector2.Normalize(diff)
            mNextDirAngle = Math.Atan2(-mNextDirVector.Y, mNextDirVector.X)
            SetVelocity(nextspeed * mNextDirVector)
        End If
        MyBase.Update(deltaTime)
    End Sub
    Public Sub SetSpeed(ByVal value As Double)
        speed = value
    End Sub
    Public Sub SetStartNode(ByRef start As Tile)
        If (start IsNot Nothing) Then mNextNode = start.GetParent()
    End Sub
    Public Function GetNextNode() As Tile
        Return mNextNode
    End Function
    Public Function GetNextDirVector() As Vector2
        Return mNextDirVector
    End Function
    Public Function GetNextDirAngle() As Double
        Return mNextDirAngle
    End Function

    Private mNextNode As Tile
    Private speed As Double
    Private mNextDirVector As Vector2
    Private mNextDirAngle As Double

End Class
